File: src/cpp/lesson_02_publisher/src/publisher_node.cpp
#include "lesson_02_publisher_cpp/publisher_node.hpp"

PublisherNode::PublisherNode(const rclcpp::NodeOptions &options)
: rclcpp::Node("lesson_02_publisher", options) {
  prefix_ = declare_parameter<std::string>("message_prefix", "Hello");
  const double period_s = declare_parameter<double>("publish_period_s", 0.5);
  period_ = std::chrono::duration_cast<std::chrono::milliseconds>(
      std::chrono::duration<double>(period_s));

  publisher_ = create_publisher<Msg>(topics::CHATTER, qos::telemetry());
  timer_ = create_wall_timer(period_, [this]() { publish_message(); });

  RCLCPP_INFO(
      get_logger(),
      "Lesson 02 publisher started (topic: %s, period: %ld ms).",
      topics::CHATTER,
      static_cast<long>(period_.count()));
}

void PublisherNode::publish_message() {
  Msg msg;
  msg.data = prefix_ + " " + std::to_string(count_++);
  publisher_->publish(msg);
}

int main(int argc, char **argv) {
  try {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<PublisherNode>();
    rclcpp::spin(node);
  } catch (const std::exception &e) {
    RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Exception in main: %s", e.what());
  }
  rclcpp::shutdown();
  return 0;
}

File: src/1_python/lesson_05_parameters/lesson_05_parameters_py/publisher_node.py
#!/usr/bin/env python3
"""
Lesson 05 Publisher Node.
Demonstrates component composition and centralized configuration.
"""
from __future__ import annotations

from typing import List, Optional

import rclpy
from rcl_interfaces.msg import SetParametersResult
from rclpy.node import Node

from lesson_interfaces.msg import MsgCount
from utils_py import qos, topics

from lesson_05_parameters_py.utils import validate_timer_period_s
from lesson_05_parameters_py.logic import TelemetryGenerator


class TelemetryPublisher:
    """
    Encapsulates publishing logic and resource ownership.
    """

    def __init__(self, node: Node) -> None:
        # 1. Instantiate Pure Logic
        self._logic = TelemetryGenerator()

        # 2. Setup ROS Resources
        topic_name = topics.telemetry(node)
        qos_profile = qos.telemetry(node)

        self._handle = node.create_publisher(MsgCount, topic_name, qos_profile)
        node.get_logger().info(f"TelemetryPublisher initialized on '{topic_name}'")

    def publish(self) -> None:
        """Construct and send the next message."""
        val = self._logic.next_value()
        
        msg = MsgCount()
        msg.count = val
        self._handle.publish(msg)


class Lesson05PublisherNode(Node):
    """
    Manages the application lifecycle, parameters, and execution scheduling.
    """

    def __init__(self) -> None:
        super().__init__("lesson_05_publisher")

        self._timer = None
        self._timer_period_s = 1.0

        # 1. Setup
        self._declare_parameters()
        self._telemetry = TelemetryPublisher(self)

        # 2. Execution (Start Timer)
        initial_period = self.get_parameter("timer_period_s").value
        self._create_or_update_timer(initial_period)

        # 3. Callbacks
        self.add_on_set_parameters_callback(self._on_param_update)

        self.get_logger().info("Lesson 05 publisher started.")

    def _declare_parameters(self) -> None:
        self.declare_parameter("timer_period_s", 1.0)

    def _on_param_update(self, parameters: List[rclpy.Parameter]) -> SetParametersResult:
        for parameter in parameters:
            if parameter.name == "timer_period_s":
                validation = validate_timer_period_s(parameter.value)
                if not validation.ok:
                     return SetParametersResult(successful=False, reason=validation.reason)
                
                self._create_or_update_timer(validation.value)

        return SetParametersResult(successful=True)

    def _create_or_update_timer(self, period_s: float) -> None:
        """
        Idempotent timer creation. Only destroys/recreates if period changes.
        """
        # Optimization: No-op if frequency matches
        if self._timer is not None and self._timer_period_s == period_s:
            return

        if self._timer is not None:
            self._timer.cancel()
            self._timer.destroy()

        self._timer_period_s = period_s
        self._timer = self.create_timer(self._timer_period_s, self._telemetry.publish)
        
        self.get_logger().info(f"Timer set to {self._timer_period_s}s")


def main(args: Optional[List[str]] = None) -> None:
    rclpy.init(args=args)
    node: Optional[Lesson05PublisherNode] = None

    try:
        node = Lesson05PublisherNode()
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    except Exception as exc:
        logger = rclpy.logging.get_logger("lesson_05_publisher")
        logger.error(f"Unhandled exception: {exc}")
    finally:
        if node is not None:
            node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == "__main__":
    main()